data_model/src/events/notification.rs

```diff
@@ -0,0 +1,290 @@
+        TriggerCompleted(TriggerCompletedEventFilter),
+    }
+
+    /// Filter [`TriggerCompletedEvent`] by
+    /// 1. if `triger_id` is some filter based on trigger id
+    /// 2. if `outcome_type` is some filter based on execution outcome (success/failure)
+    /// 3. if both fields are none accept every event of this type
+    #[derive(
+        Debug,
+        Clone,
+        Constructor,
+        Getters,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    #[ffi_type]
+    #[getset(get = "pub")]
+    #[non_exhaustive]
+    pub struct TriggerCompletedEventFilter {
+        trigger_id: Option<TriggerId>,
+        outcome_type: Option<TriggerCompletedOutcomeType>,
+    }
+
+    /// Type of outcome to filter on
+    #[derive(
+        Debug,
+        Clone,
+        PartialEq,
+        Eq,
+        PartialOrd,
+        Ord,
+        Decode,
+        Encode,
+        Deserialize,
+        Serialize,
+        IntoSchema,
+    )]
+    #[repr(u8)]
+    #[ffi_type]
+    pub enum TriggerCompletedOutcomeType {
+        Success,
+        Failure,
+    }
+}
+
+#[cfg(feature = "transparent_api")]
+impl super::Filter for NotificationEventFilter {
+    type Event = NotificationEvent;
+
+    /// Check if `self` accepts the `event`.
+    #[inline]
+    fn matches(&self, event: &Self::Event) -> bool {
+        match (self, event) {
+            (Self::AcceptAll, _) => true,
+            (Self::TriggerCompleted(filter), NotificationEvent::TriggerCompleted(event)) => {
+                filter.matches(event)
+            }
+        }
+    }
+}
+
+#[cfg(feature = "transparent_api")]
+impl super::Filter for TriggerCompletedEventFilter {
+    type Event = TriggerCompletedEvent;
+
+    /// Check if `self` accepts the `event`.
+    #[inline]
+    fn matches(&self, event: &Self::Event) -> bool {
+        if let Some(trigger_id) = self.trigger_id() {
+            if trigger_id != event.trigger_id() {
+                return false;
+            }
+        }
+
+        if let Some(outcome_type) = self.outcome_type() {
+            match (outcome_type, event.outcome()) {
+                (TriggerCompletedOutcomeType::Success, TriggerCompletedOutcome::Failure(_))
+                | (TriggerCompletedOutcomeType::Failure, TriggerCompletedOutcome::Success) => {
+                    return false
+                }
+                _ => {}
+            }
+        }
+
+        true
+    }
+}
+
+/// Exports common structs and enums from this module.
+pub mod prelude {
+    pub use super::{
+        NotificationEvent, NotificationEventFilter, TriggerCompletedEvent,
+        TriggerCompletedEventFilter, TriggerCompletedOutcome, TriggerCompletedOutcomeType,
+    };
+}
+
+#[cfg(test)]
+#[cfg(feature = "transparent_api")]
+mod tests {
+    use super::*;
+    use crate::events::Filter;
+
+    #[test]
+    fn trigger_completed_events_filter() {
```

Comment:
Shouldn't it be different tests instead of just a big one?

Reply:
Not sure, we will have to repeat initialization steps multiple times + `assert`s point to exact location where error is happened.

Reply:
I don't get the idea about asserts, but agree about multiple initialization. Not sure too. Maybe we need more opinions here

Reply:
> I don't get the idea about asserts

I mean that purpose of multiple tests is to localize problems quicker. In case of assert we immediately get LOC where error happened and so we localize bug.

Reply:
Big fan of the asserts.